src/openzeppelin/token/erc1155/library.cairo

```diff
@@ -0,0 +1,786 @@
+        with_attr error_message("ERC1155: cannot call transfer from the zero address"):
+            assert_not_zero(caller)
+        end
+        with_attr error_message("ERC1155: transfer caller is not owner nor approved"):
+            assert_owner_or_approved(from_)
+        end
+        _safe_batch_transfer_from(from_, to, ids_len, ids, amounts_len, amounts, data_len, data)
+        return ()
+    end
+
+    #
+    # Internals
+    #
+
+    func _safe_transfer_from{
+            syscall_ptr: felt*,
+            pedersen_ptr: HashBuiltin*,
+            range_check_ptr
+        }(
+            from_: felt,
+            to: felt,
+            id: Uint256,
+            amount: Uint256,
+            data_len: felt,
+            data: felt*
+        ):
+        alloc_locals
+        # Check args
+        with_attr error_message("ERC1155: transfer to the zero address"):
+            assert_not_zero(to)
+        end
+        with_attr error_message("ERC1155: id is not a valid Uint256"):
+            uint256_check(id)
+        end
+        with_attr error_message("ERC1155: amount is not a valid Uint256"):
+            uint256_check(amount)
+        end
+
+        # Deduct from sender
+        let (from_balance: Uint256) = ERC1155_balances.read(id, from_)
+        with_attr error_message("ERC1155: insufficient balance for transfer"):
+            let (new_balance: Uint256) = SafeUint256.sub_le(from_balance, amount)
+        end
+        ERC1155_balances.write(id, from_, new_balance)
+
+        # Add to receiver
+        let (to_balance: Uint256) = ERC1155_balances.read(id, to)
+        with_attr error_message("ERC1155: balance overflow"):
+            let (new_balance: Uint256) = SafeUint256.add(to_balance, amount)
+        end
+        ERC1155_balances.write(id, to, new_balance)
+
+        # Emit events and check
+        let (operator) = get_caller_address()
+        TransferSingle.emit(
+            operator,
+            from_,
+            to,
+            id,
+            amount
+        )
+        _do_safe_transfer_acceptance_check(
+            operator,
+            from_,
+            to,
+            id,
+            amount,
+            data_len,
+            data
+        )
+        return ()
+    end
+
+    func _safe_batch_transfer_from{
+            syscall_ptr: felt*,
+            pedersen_ptr: HashBuiltin*,
+            range_check_ptr
+        }(
+            from_: felt,
+            to: felt,
+            ids_len: felt,
+            ids: Uint256*,
+            amounts_len: felt,
+            amounts: Uint256*,
+            data_len: felt,
+            data: felt*
+        ):
+        alloc_locals
+        # Check args
+        with_attr error_message("ERC1155: transfer to the zero address"):
+            assert_not_zero(to)
+        end
+        with_attr error_message("ERC1155: ids and amounts length mismatch"):
+            assert ids_len = amounts_len
+        end
+        # Recursive call
+        let len = ids_len
```

Comment:
The role of `len` here and elsewhere is to make it explicit that it is the length of both `ids` and `amounts`.  Using kwargs would also make this explicit, but this doesn't add any steps either way. We can drop it outright (since it _is_ explicit in the signature of `safe_batch_transfer_from_iter`) or switch to kwargs (i.e. `len=ids_len`) if you prefer?